### ISSUE TITLE
Figure out what to do with webhook authz cache in large clusters

### ISSUE THREAD

**Issue by mikedanese** [kubernetes/kubernetes#85855]
Some options are:

* Increasing LRU cache size
* Migrate to expiring cache
* Share authorization lookups

/sig auth

**Comment by mikedanese** [comment]
@kubernetes/sig-auth-feature-requests

**Comment by wojtek-t** [comment]
Sharing authentication looks is definitely a good idea, but it may not be enough.

Regarding the first two options:

> * Increasing LRU cache size
> * Migrate to expiring cache

I'm not sure migrating to expiring cache is safe enough. While probably in 99,99% of cases, it would solve all the problems, in some pathological scenarios, it may end-up with huge increase of memory and may in theory result in OOMs, which would be really bad user experience.

I think that LRU cache is actually good enough, we just need to tweak its size.
So let me share a couple thoughts:
- I think that number of nodes in the cluster would actually be a pretty good proxy for the size (usually requests from nodes are the main contributor to the problem)
- we already have that data in apiserver - we even expose its approximation as prometheus metric:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1353

So maybe the easiest way to do it to:
- periodically check cluster size
- resize LRU cache based on that:
(a) while the current version of that doesn't support Resize() operation
(b) the head of that repo actually does: https://github.com/hashicorp/golang-lru/blob/master/lru.go#L97 (it was added 2 years ago)

[Exercising /metrics to get this data is probably not the best option, but it should be possible to pipe this through the code]

@mikedanese - thoughts?

**Comment by liggitt** [comment]
> * I think that number of nodes in the cluster would actually be a pretty good proxy for the size (usually requests from nodes are the main contributor to the problem)

I'm not in favor of pinning sizing for generic operations like authn/authz to the number of nodes in the cluster. With CRDs, servers handle more use cases, and the generic mechanisms like admission/authn/authz need to scale reliably apart from number of nodes.

**Comment by wojtek-t** [comment]
> With CRDs, servers handle more use cases, and the generic mechanisms like admission/authn/authz need to scale reliably apart from number of nodes.

I agree with that. But it's order of magniture better than not scale it at all (which is the current state).

**Comment by mikedanese** [comment]
How much bigger do you need the LRU cache to be to hit node counts passed 10k? Other options are:
* Don't scale the size, just adjust it to satisfy the highest counts we need to hit. This works if the cache is small compared to the overall memory usage of the master.
* Scale it based on another metric like available memory on the master node. 
* Plumb it as a flag and let the cluster operator make the decision.

**Comment by mikedanese** [comment]
Also note that this issue is only for the webhook authorizer. Nodes should be authorized in general by the node authorizer, and the results aren't cached.

**Comment by wojtek-t** [comment]
> Don't scale the size, just adjust it to satisfy the highest counts we need to hit. This works if the cache is small compared to the overall memory usage of the master.

@jkaniuk  - I think you were doing some experiments around that. Do you have any numbers?

> Scale it based on another metric like available memory on the master node.

We did a terrible hack for that in GCE scripts:
https://github.com/kubernetes/kubernetes/blob/9719ffe4ae644b7197acb82e69309bbae0e4c70a/cmd/kube-apiserver/app/server.go#L673
https://github.com/kubernetes/kubernetes/blob/9d0a32caf8bf9f2fdc31ee586938bdbaba64fab4/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go#L179
https://github.com/kubernetes/kubernetes/blob/4c50ee993c82c6852eb3b3aa8dfa8ecc4bcfe330/cluster/gce/gci/configure-kubeapiserver.sh#L124

If we do that in nicer way, it may work.

> Plumb it as a flag and let the cluster operator make the decision.

I would prefer not to do that - it's too manual imho.

I'm generally open on any other metrics too - node count was only one suggestion. But I would prefer not to move the necessity to think about it on the operator.

**Comment by mikedanese** [comment]
Let's raise the priority once this becomes an issue.
